docs/source/argument_check.rst

```diff
@@ -0,0 +1,636 @@
+
+- If the rocBLAS kernel requires no temporary device memory, the macro ``RETURN_ZERO_DEVICE_MEMORY_SIZE_IF_QUERIED(handle)`` can be called after checking that ``handle != nullptr``.
+
+- If the rocBLAS kernel requires temporary device memory, then it should be set, and the kernel returned, by calling ``return handle->set_optimal_device_memory_size(size...)``, where ``size...`` is a list of one or more sizes for different sub-problems. The sizes are rounded up and added.
+
+Logging
+-------
+
+- There is logging before a quick-return-success or error-return, except:
+
+  - when ``handle == nullptr``, return ``rocblas_status_invalid_handle`` 
+  - when ``handle->is_device_memory_size_query()`` returns ``true``
+
+- Vectors and matrices are logged with their addresses, and are always on device memory.
+
+- Scalar values in device memory are logged as their addresses. Scalar values in host memory are logged as their values, with a ``nullptr`` logged as ``NaN`` (``std::numeric_limits<T>::quiet_NaN()``).
+
+rocBLAS control flow:
+=====================
+
+1. If ``handle == nullptr``, return ``rocblas_status_invalid_handle``.
+
+2. If the function does not require temporary device memory, call the macro ``RETURN_ZERO_DEVICE_MEMORY_SIZE_IF_QUERIED(handle);``.
+
+3. If the function requires temporary device memory, and ``handle->is_device_memory_size_query()`` is ``true``, then validate any pointers and arguments required to determine the optimal size of temporary device memory, returning ``rocblas_status_invalid_pointer`` or ``rocblas_status_invalid_size`` if the arguments are invalid, and otherwise ``return handle->set_optimal_device_memory_size(size...);``, where ``size...`` is a list of one or more sizes of temporary buffers, which are allocated with ``handle->device_malloc(size...)`` later.
+
+4. Perform logging if enabled, taking care not to dereference ``nullptr`` arguments.
+
+5. Check for invalid sizes. Return ``rocblas_status_invalid_size`` if size arguments are invalid.
+
+3. Check for unsupported enum value. Return ``rocblas_status_invalid_value`` if enum value is invalid.
+
+6. Return ``rocblas_status_invalid_pointer`` if any pointers used to determine quick return conditions are NULL.
+
+7. If quick return conditions are met:
+
+   - if there is no return value
+
+     - Return ``rocblas_status_success``
+
+   - If there is a return value
+
+     - If the return value pointer argument is nullptr, return ``rocblas_status_invalid_pointer``
+
+     - Else, return ``rocblas_status_success``
+
+9. Check for NULL pointer arguments not already covered by #3. Return ``rocblas_status_invalid_pointer`` if argument pointers are NULL.
+
+10. (Optional.) Allocate device memory, returning ``rocblas_status_memory_error`` if the allocation fails.
+
+11. After completing the calculation, return ``rocblas_status_success``, assuming no errors in HIP calls or other errors in the calculation.
+
```

Comment:
Not after completing, "After launching the calculation".   Only know that the calculations were successfully launched asynchronously.